hpws.c

```diff
@@ -521,22 +534,27 @@ int main(int argc, char **argv)
                             int msg_size = 2;
                             *(uint8_t *)&visa_msg_buf = UDP_MSG_CHALLENGE;
 
-                            if (is_too_old)
-                                *(uint8_t *)((uint8_t *)&visa_msg_buf + 1) = VISA_MSG_EXPIRED;
+                            if (err_no>0){
+                                *(uint8_t *)((uint8_t *)&visa_msg_buf + 2) = err_no;
+                            }
                             // Check if the visa data valid.
                             else if (memcmp(((unsigned char *)&visa_msg_buf + 5), &visa_token, sizeof(visa_token)) != 0)
                                 *(uint8_t *)((uint8_t *)&visa_msg_buf + 1) = VISA_MSG_FAILED;
                             else
                             {
-                                const unsigned char challenge[CHALLENGE_SIZE] = {1};
-
-                                *(uint8_t *)&visa_msg_buf = UDP_MSG_CHALLENGE;
-                                memcpy(((unsigned char *)&visa_msg_buf + 1), &challenge, CHALLENGE_SIZE);
+                                *(uint8_t *)((uint8_t *)&visa_msg_buf + 1) = VISA_MSG_ACCEPTED;
+                                
+                                // Seed the random number generator with the current time
+                                srand(time(NULL));
+                                uint32_t random_number = rand();
+                                const unsigned char challenge[CHALLENGE_SIZE];
+                                generate_hashed_random_number(random_number, challenge);
```

Comment:
Instead declaring challenge buffer we can call `generate_hashed_random_number(random_number, ((unsigned char *)&visa_msg_buf + 2));`
Then we can remove the memcpy as well